### ISSUE TITLE
Centered layout is active on start if vscode was closed in zenmode

### ISSUE THREAD

**Issue by SrTobi** [Microsoft/vscode#45610]
Issue Type: <b>Bug</b>

-----

1. Open vscode
2. Make sure you are **not** in zenmode and the layout is **not** centered.
3. Make sure the option `zenMode.centerLayout` is `true`
4. Toggle Zen mode, so that you are in zenmode (the layout should be centered)
5. Exit the editor without exiting the zenmode (e.g. `alt+F4` on windows)
6. Restart vscode

Now the centered layout is still activated outside of the zen mode which it should not be

------

VS Code version: Code 1.21.0 (9a199d77c82fcb82f39c68bb33c614af01c111ba, 2018-03-07T11:01:43.521Z)
OS version: Linux x64 4.15.7-1-ARCH

Reproduces without extensions

**Comment by SrTobi** [comment]
Added PR #45611

**Comment by isidorn** [comment]
Thanks for your PR, after reviewing this I choose to take a different approach captured in this commit d3c7be9f1a.

Basically the issue is that when the user goes out of zenMode he stays in centeredMode. They should be coupled and that commit fixes that in the same way that is done for the sidebar and for the panel. When exiting zen mode the visibility of those parts changes only if zen mode changed them.
I believe the commit does it in a cleaner and more general way than the PR thus I choose that one. Hope that makes sense.

**Comment by SrTobi** [comment]
ok, yeah I see what you did there and that's fine, I think. However, that fixes in no way the problem I described above (I tested your commit). The problem is that when you exit vscode the zenmode is not deactivated (toggled out of) and so neither is the centered layout. But instead the zenmode is implicitly exited because it simply doesn't save that zenmode was active at all. That doesn't apply to the centered layout, though, and so when you start the editor again the zenmode is not active anymore, but the centered layout is which is not what you'd expect because you want the centered layout in zenmode but not in normal mode.

**Comment by SrTobi** [comment]
Uiii... no that you mentioned the side panel I wanted to find out how the problem I described above is avoided for them... and it turns out it simply isn't :smile:  It has the same problem and so does the window size! The problem is that an exit while in zenmode doesn't quit the zenmode properly (i.e. it doesn't toggle out of it). The cleanest solution might be to check — when vscode shuts down — if the zenmode is active and if so quit it properly... which is exactly the logic of the zenmode being not active at startup again.

**Comment by isidorn** [comment]
@SrTobi for me exiting vscode and starting it again properly exits zen mode.

Please make sure to not have the "zenMode.restore": true setting which is false by default.
Also make sure to exit vscode, and not only reload.

If the zen mode still does not exit for you then it sounds like a bug and an issue with repro steps would be great :)

**Comment by SrTobi** [comment]
No `zenMode.restore` is not the problem. And I did exit vscode completely.. no reload...

Ok, I try to be more specific:

1. Open vscode
2. Make sure you are **not** in zenmode and the layout is **not** centered.
3. Make sure the option `zenMode.centerLayout` is true (which is the default) and `zenMode.restore` is false (which is also the default)
4. Toggle Zen mode, so that you are in zenmode
    - Now the layout is centered as expected, because `toggleZenMode` calls `centerEditorLayout(true,...)`
    - Here is now the problem: `centerEditorLayout` stores `true` in `Workbench.centeredEditorLayoutActiveStorageKey`
5. Exit the editor without exiting the zenmode. You do that by pressing `Ctrl+Q` or pressing the `Alt` key and select `File -> Exit` (all while in zenmode!)
    - Now, the editor is not open anymore
    - Note, however, that `Workbench.centeredEditorLayoutActiveStorageKey` is still `true` in the storage, because neither `toggleZenMode` nor `centerEditorLayout` is called
6. Start vscode
     - Now, vscode is open again and the centered layout is activated despite the fact that the zenmode is not active
     - This happens because `Workbench.centeredEditorLayoutActiveStorageKey` still holds `true` and when vscode opens it loads that value and activates the centered layout
     - That is not the expected behavior as one would expect the centered layout not to be active, because exiting vscode while in zenmode should close the zenmode properly
     - This is not only the case for the centered layout but also for the sidebar's visibility and the window size (as described in my previous comment)

What I mean with closing the zenmode properly is that all depended settings (like centered layout, window size, sidebar visibility) should have the same states as before starting the zenmode, which is normally the case because `toggleZenMode` restores them to their original states when toggling out of the zenmode

When `Ctrl+Q` is pressed (while in zenmode), however, `toggleZenMode` not called again and so the states are not restored!

**Comment by isidorn** [comment]
@SrTobi thanks those are good steps and I could reproduce also on my os x.
I have tackled this issue via this commit d67d53d1db

**Comment by SrTobi** [comment]
:+1: thanks, lgtm

**Comment by isidorn** [comment]
Adding verified label since I guess you tried it out from source.
If for some reason it still does not work in tomorrow's insiders let me know, thanks

**Comment by SrTobi** [comment]
@isidorn hi, I tested this a little further and while it works fine for `zenmode.restore = false` it does not correctly preserve the states if `zenmode.restore` is true. the problem was that `wasSideBarVisible` and the likes did not have the correct bool on restore. I made a PR #45773

Now, though, I have encountered another highly unusual problem where a `vscode:leaveFullScreen` ipc call is generated out of nothing which messes the whole workbench up... I have to investigate that further before I can report... I will open a new issue for that, however.
